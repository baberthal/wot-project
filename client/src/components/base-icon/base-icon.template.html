<span class="icon" :class="[iconType, size]">
  <i :class="[pack, iconClass, effectiveIconSize, customClass]" />
</span>
