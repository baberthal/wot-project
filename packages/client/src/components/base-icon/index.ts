//===- components/base-icon/index.ts - Re-Exports --------------------------===//
//
// Copyright (c) 2018 J. Morgan Lieberthal
// Licensed under the MIT License
//
//===-----------------------------------------------------------------------===//

export { default } from "./base-icon.component";
