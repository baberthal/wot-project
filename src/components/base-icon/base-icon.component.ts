//===- base-icon.component.ts - Base Icon Component ------------------------===//
//
// Copyright (c) 2018 J. Morgan Lieberthal
// Licensed under the MIT License
//
//===-----------------------------------------------------------------------===//

import { Component, Prop, Vue } from "@app/core";
import template from "./base-icon.template.html";

@Component({ template })
export class BaseIcon extends Vue {
  @Prop([String, Object]) type!: string | object;
  @Prop(String) icon!: string;
  @Prop(String) size!: string;
  @Prop(String) customSize!: string;
  @Prop(String) customClass!: string;

  get iconClass(): string {
    return `${this.pack}-${this.icon}`;
  }

  get pack(): string {
    return "mdi";
  }

  get iconType(): string | undefined {
    if (!this.type) return;

    let splitType: string[] = [];

    if (typeof this.type === "string") {
      splitType = this.type.split("-");
    } else {
      for (let key in this.type) {
        if ((this.type as any)[key]) {
          splitType = key.split("-");
          break;
        }
      }
    }

    if (splitType.length <= 1) return;

    return `has-text-${splitType[1]}`;
  }

  get effectiveIconSize(): string | undefined {
    return this.customSize || this.customSizeByPack;
  }

  get customSizeByPack(): string | undefined {
    switch (this.size) {
      case "is-small":
        return;
      case "is-medium":
        return "mdi-36px";
      case "is-large":
        return "mdi-48px";
      default:
        return "mdi-24px";
    }
  }
}
