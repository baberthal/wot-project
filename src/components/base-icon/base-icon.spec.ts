//===- components/base/base-icon/base-icon.spec.ts - Base Icon Spec --------===//
//
// Copyright (c) 2018 J. Morgan Lieberthal
// Licensed under the MIT License
//
//===-----------------------------------------------------------------------===//

import { shallowMount } from "@vue/test-utils";
import { BaseIcon } from "./base-icon.component";

describe("BaseIcon", () => {
  it("renders correctly", () => {
    const wrapper = shallowMount(BaseIcon);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("is a vue instance", () => {
    const wrapper = shallowMount(BaseIcon);

    expect(wrapper.name()).toBe("BaseIcon");
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it("render icon when icon property is passed", () => {
    const wrapper = shallowMount(BaseIcon, {
      propsData: { icon: "eye" }
    });

    expect(wrapper.classes()).toContain("icon");
    expect(wrapper.find("i").classes()).toEqual(
      expect.arrayContaining(["mdi", "mdi-eye", "mdi-24px"])
    );
  });

  it("render a colored icon when type is passed", () => {
    const wrapper = shallowMount(BaseIcon, {
      propsData: {
        icon: "eye",
        type: "is-primary"
      }
    });

    expect(wrapper.classes()).toContain("has-text-primary");
  });

  it("display size when size propery is passed", () => {
    const wrapper = shallowMount(BaseIcon, {
      propsData: {
        icon: "eye"
      }
    });

    expect(wrapper.find("i").classes()).toContain("mdi-24px");
    wrapper.setProps({ size: "is-small" });
    expect(wrapper.find("i").classes()).toEqual(
      expect.arrayContaining(["mdi", "mdi-eye"])
    );
    wrapper.setProps({ size: "is-medium" });
    expect(wrapper.find("i").classes()).toContain("mdi-36px");
    wrapper.setProps({ size: "is-large" });
    expect(wrapper.find("i").classes()).toContain("mdi-48px");
  });

  it("render custom classes when customClass property is passed", () => {
    const wrapper = shallowMount(BaseIcon, {
      propsData: {
        icon: "eye",
        customClass: "foo-bar"
      }
    });

    expect(wrapper.find("i").classes()).toContain("foo-bar");
  });
});
