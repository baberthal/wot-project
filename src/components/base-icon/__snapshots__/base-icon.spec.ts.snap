// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseIcon renders correctly 1`] = `<span class="icon"><i class="mdi mdi-undefined mdi-24px"></i></span>`;
